### PR TITLE
docs: add chainlink-integration guide

### LINKED ISSUE #379: Docs to answer: how to connect agoric with chainlink
Write up an overview answering the question "how to connect agoric with chainlink".

We should mention testing with mocks first.

### PR DESCRIPTION
**Pull request by michaelfig**
Closes #379
Closes Agoric/dapp-oracle#9

### CODE REVIEW

File: main/chainlink-integration.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
+# Chainlink Integration
+
+This document explains how you can consume [Chainlink
```
Comment (tyg) [inline main/chainlink-integration.md:3]:
```suggestion
This document explains how to consume [Chainlink
```

"you can consume" reads like Chainlink oracles are a crunchy snack you can chow down on. : -)

File: main/chainlink-integration.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+oracles](https://chain.link) when unit testing, integration testing, and
+actually on-chain.
+
+## Overview
+
+Using Chainlink on Agoric provides two main features:
+- price feeds exposed on Agoric via the on-chain `home.priceAuthority`, an
+  officially-sponsored [price authority](/zoe/guide/price-authority.md) built
+  from aggregating several Chainlink nodes.
```
Comment (tyg) [inline main/chainlink-integration.md:12]:
```suggestion
- Price feeds are exposed on Agoric via the on-chain `home.priceAuthority`. This is an
  officially-sponsored [price authority](/zoe/guide/price-authority.md) built
  from aggregating several Chainlink nodes.
```

File: main/chainlink-integration.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+- [Chainlink's Any API](https://docs.chain.link/docs/request-and-receive-data)
+  to initiate a job on a single oracle and return its results
```
Comment (tyg) [inline main/chainlink-integration.md:14]: This doesn't sync with the intro to it, which is "Using Chainlink provides two main features:".  Maybe something like change "to initiate" to "can initiiate"

File: main/chainlink-integration.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+These features have been tested with [actual Chainlink oracle
```
Comment (tyg) [inline main/chainlink-integration.md:16]:
```suggestion
We have tested these features with [actual Chainlink oracle
```

File: main/chainlink-integration.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+software](https://github.com/Agoric/dapp-oracle/blob/main/chainlink-agoric/README.md).
+
+However, it is important to note that Chainlink is still in the process of
+setting up an incentivized testnet for established Chainlink node operators to
+connect to Agoric.
```
Comment (tyg) [inline main/chainlink-integration.md:21]:
```suggestion
**Note**: Chainlink has not yet finished setting up an incentivized testnet for established Chainlink node operators to connect to Agoric.
```

Maybe add an ETA?

File: main/chainlink-integration.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+## Price Authority
+
+To test your contract against a locally-simulated price authority, just follow
+the instructions in [the Price Authority API](/zoe/guide/price-authority.md).
+
+If you want to use the curated on-chain price authority, you can find it at
+`home.priceAuthority`.  Here is a sample of getting a quote for selling `30
+Testnet.$LINK` in `Testnet.$USD`:
```
Comment (tyg) [inline main/chainlink-integration.md:30]:
```suggestion
To use the curated on-chain price authority, see `home.priceAuthority`.  For example, to get a quote for selling `30
Testnet.$LINK` in `Testnet.$USD`:
```

File: main/chainlink-integration.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+const { quoteAmount: { value: [{ amountOut: usdAmount, timestamp }] } } = await E(home.priceAuthority).quoteGiven(linkAmount, usdBrand);
+const linkValue = linkAmount.value / 10 ** linkDecimals;
+const usdValue = usdAmount.value / 10 ** usdDecimals;
+console.log(linkValue, 'Testnet.$LINK trades for', usdValue, 'Testnet.$USD');
+// 30 Testnet.$LINK trades for 12 Testnet.$USD
+```
+
+## Any API
+
+In order to use Chainlink's Any API, you will need to obtain an instance of the
```
Comment (tyg) [inline main/chainlink-integration.md:52]:
```suggestion
To use Chainlink's Any API, you need to get an instance of the
```

File: main/chainlink-integration.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+```js
+{
+  jobId: <Chainlink JobId>,
+  params: { ...<job parameters> }
+}
+```
+
+The string `result` data that the oracle node returns will be sent as the
+response to your query.
```
Comment (tyg) [inline main/chainlink-integration.md:64]:
```suggestion
The string `result` data the oracle node returns is sent as the
response to your query.
```

File: main/zoe/guide/contracts/oracle.md [hunk 9]
```diff
@@ -1,12 +1,22 @@
...
-This contract allows other contracts or users to make queries to an
-external oracle. The queries can be free, or require a fee. 
+**NOTE: You almost certainly do not want to use this contract directly.
+Instead, please read the [Chainlink integration
+documentation](/chainlink-integration.md)**
+
+This contract allows other contracts or users to make single queries to a generic
+oracle. The queries can be free, or require a fee.  This provides a very
+low-level API to issue single queries that are answered by an individual
+off-chain oracle node.
```
Comment (tyg) [inline main/zoe/guide/contracts/oracle.md:15]:
```suggestion
This contract lets other contracts or users make single free or fee-based queries to a generic
oracle.  This provides a very
low-level API to issue single queries that an individual off-chain oracle node answers.
```

File: main/chainlink-integration.md [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+query of the form:
+
+```js
+{
+  jobId: <Chainlink JobId>,
+  params: { ...<job parameters> }
+}
+```
+
+The string `result` data the oracle node returns is sent as the
```
Comment (tyg) [inline main/chainlink-integration.md:60]:
```suggestion
The string `result` data the oracle node returns is sent as the
```
I'm not quite following what you mean by "The string 'result' data". Is the returned data just 'result'? Is it data prefaced by 'result'? In general, I'd like to change the sentence, depending on what those first four words mean, to something like:
"The oracle node returns <whatever>. This is sent as your query's response."

File: main/chainlink-integration.md [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+response to your query.
+
+You can test these queries against a locally-running Chainlink node that you
+control by following [the Chainlink integration
+instructions](https://github.com/Agoric/dapp-oracle/blob/main/chainlink-agoric/README.md).
```
Comment (tyg) [inline main/chainlink-integration.md:65]:
```suggestion
You can test these queries against a locally-running Chainlink node. You
control the node by following [the Chainlink integration
instructions](https://github.com/Agoric/dapp-oracle/blob/main/chainlink-agoric/README.md).
```

File: main/chainlink-integration.md [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+If you want to use a more limited local node that emulates just a portion of the
+Chainlink API without having to run Docker containers, you can use the [local
+(mock) builtin
+oracle](https://github.com/agoric/dapp-oracle#running-a-local-builtin-oracle).
```
Comment (tyg) [inline main/chainlink-integration.md:70]:
```suggestion
There is a more limited local node that emulates part of the
Chainlink API without having to run Docker containers.  This is the [local
(mock) builtin
oracle](https://github.com/agoric/dapp-oracle#running-a-local-builtin-oracle).
```

File: main/chainlink-integration.md (hunk 10 above)
Comment (michaelfig) [reply main/chainlink-integration.md:60]: Thanks, I've clarified it to: "The oracle node returns its result, which is conventionally a string. This is sent as your query's reply."

File: main/chainlink-integration.md (hunk 11 above)
Comment (michaelfig) [reply main/chainlink-integration.md:65]:
```suggestion
You can test these queries against a locally-running Chainlink node that you
control.  Follow [the Chainlink integration
instructions](https://github.com/Agoric/dapp-oracle/blob/main/chainlink-agoric/README.md) to set it up.
```

**Review by tyg — APPROVED** [review]

File: main/zoe/guide/contracts/oracle.md [hunk 13]
```diff
@@ -1,12 +1,21 @@
...
-external oracle. The queries can be free, or require a fee. 
+**NOTE: You almost certainly do not want to use this contract directly.
+Instead, please read the [Chainlink integration
+documentation](/chainlink-integration.md)**
+
+This contract lets other contracts or users make single free or fee-based queries to a generic
+oracle.  This provides a very
+low-level API to issue single queries that an individual off-chain oracle node answers.
+
+Relying on just a single node can be both expensive and risky. Instead, use the
```
Comment (katelynsills) [inline main/zoe/guide/contracts/oracle.md:16]: I think Chainlink already has aggregators on their side.

**Comment by katelynsills** [comment]
Sorry, somehow I missed that I was added as a reviewer

File: main/zoe/guide/contracts/oracle.md (hunk 13 above)
Comment (michaelfig) [reply main/zoe/guide/contracts/oracle.md:16]: > I think Chainlink already has aggregators on their side.

They don't on the Agoric chain.  That's implemented in Solidity on Ethereum (and in the higher-level contracts I allude to in Agoric), not in the oracle nodes themselves.

Comment (katelynsills) [reply main/zoe/guide/contracts/oracle.md:16]: Ah, gotcha. Can we explain this further? I think folks may have the mental model that I have.

Comment (michaelfig) [reply main/zoe/guide/contracts/oracle.md:16]: > Can we explain this further? I think folks may have the mental model that I have.

Yes, I'll do that.

**Pull request merged**